mne_bids/write.py

```diff
@@ -1185,7 +1185,11 @@ def write_raw_bids(raw, bids_path, events_data=None,
                 'https://github.com/mne-tools/mne-bids/issues')
         raise ValueError(msg)
 
-    datatype = _handle_datatype(raw)
+    if bids_path.datatype is None:
+        datatype = _handle_datatype(raw)
+    else:
+        datatype = bids_path.datatype
+        _check_datatype(raw, datatype)
```

Comment:
I would do this in the _handle_datatype function

`datatype = _handle_datatype(raw, bids_path.datatype)`

then you call _check_datatype in _handle_datatype and your return directly the 2nd parameter if valid.